### ISSUE TITLE
feat: confidence calibration — reduce false positives for external root causes

### ISSUE THREAD

**Issue by kamilpajak** [kamilpajak/heisenberg#49]
## Problem

When analyzing E2E test failures, Heisenberg consistently reports **95-100% confidence** attributing failures to code changes in the PR, even when the actual root cause is **external to the repository** (external configuration tools, upstream service outages, environment misconfigurations).

### Example

Azure DevOps builds 176427, 176310, 176286, 176372 all fail on E2E tests with the same pattern:
- All tests fail with identical JS TypeError / HTTP 400 errors
- LLM sees symptoms in logs + code changes in PR diff
- LLM concludes: "regression in production code" with 95% confidence
- **Reality**: external `pathConfiguration` managed by a tool outside the repository broke API routing. Code changes are unrelated.

### Root cause of the problem

The LLM operates under a **closed-world assumption** — it assumes the root cause must exist within the visible context (PR diff + logs). When it sees symptoms + code changes, it forces a correlation, producing high-confidence false positives.

### Industry approaches (Perplexity Deep Research, 2026-04-03)

**Cross-branch baseline comparison** (strongest signal):
- Datadog CI Visibility, Launchable, Google TAP all implement this
- If `main` branch also fails with the same error → exonerate the PR
- Hard rule: `confidence *= 0.1`

**Blast radius analysis** (widely implemented):
- If >25% of tests fail with identical error signature → likely infrastructure, not code regression
- 1-3 specific tests failing → likely code regression
- Tests from unrelated modules failing identically → infrastructure

**Diff-fault intersection** (Launchable: ~30% false positive reduction):
- Compare files changed in PR with files referenced in stack traces
- Zero overlap → `confidence *= 0.25`
- Limitation: catches ~60% of environmental failures (misses transitive dependencies)

**LLM confidence calibration** (academic research):
- Chain-of-thought alternative hypotheses (Kadavath et al. 2022): reduces overconfidence by 15-40%
- Evidence density scoring: direct evidence (stack trace → diff line) vs correlational (failure after change)
- Bayesian adjustment: use `P(code_caused) ≈ 20-40%` as prior, not 95%

### Expert analysis (zen code review, 2026-04-03)

**Categorization-first approach**: force the LLM to classify the failure domain before diagnosing:
- `CODE_REGRESSION` — logic error in the diff
- `INFRASTRUCTURE` — networking, 4xx/5xx, missing env vars
- `EXTERNAL_DEPENDENCY` — external config, upstream API changes
- `TEST_FLAKINESS` — race conditions, nondeterministic timeouts

**Confidence ceiling rules**:
- Stack trace references a file in the PR diff → high confidence allowed
- Stack trace references a component that interacts with diff → medium (cap 50-79%)
- Generic error (400, undefined) + no diff overlap → low (cap <50%)

### Production heuristics for false positive detection

| Signal | Inference | Confidence multiplier |
|--------|-----------|----------------------|
| >25% tests fail with same error | Infrastructure | × 0.3 |
| No files in common (diff vs stack trace) | External cause | × 0.25 |
| Same error in 2+ unrelated PRs | Infrastructure | × 0.15 |
| HTTP 5xx / connection refused in logs | Infra probable | × 0.5 |
| HTTP 4xx in logs | Could be code or config | × 0.8 |
| Test has >15% historical flake rate | Flaky test | × (1 - flake_rate) |
| Main branch also fails with identical error | Exonerate PR | × 0.1 |

### Recommended confidence adjustment pipeline

```
initial_confidence = llm_confidence  # e.g., 95%

# Apply blast radius
if error_spread > 0.25:
    initial_confidence *= 0.3

# Apply diff-fault intersection
if no_files_in_common(diff, stack_trace):
    initial_confidence *= 0.25

# Apply flake history
initial_confidence *= (1 - historical_flake_rate)

# Apply baseline branch comparison
if main_branch_has_identical_error:
    initial_confidence *= 0.1

final_confidence = max(initial_confidence, 5)
```

For the observed case (95% LLM output, all E2E fail identically, no diff overlap):
**95% × 0.3 × 0.25 = ~7%** — correctly downranked.

### Phase 1: System prompt improvement (low effort, high impact)
- Add categorization step to system prompt (CODE vs INFRA vs EXTERNAL vs FLAKE)
- Add explicit instruction: "Do not assume the PR diff caused the failure"
- Add: "If >50% of tests fail with same error, consider infrastructure cause"

### Phase 2: Blast radius heuristic (low effort)
- After clustering, calculate `error_spread = cluster_size / total_failed`
- If `error_spread > 0.25`, inject context flag into prompt
- Post-process: cap confidence if category is infra + high spread

### Phase 3: Diff-fault intersection (medium effort)
- Compare `get_pr_diff` changed files with files in stack traces
- Calculate intersection score
- Inject as context flag + use for post-processing confidence cap

### Phase 4: Baseline branch comparison (high effort, deferred)
- Requires additional API calls to check main branch status
- Most impactful but most complex — defer until Phases 1-3 are validated

## References

- Kadavath et al. (2022): "Language Models (Mostly) Know What They Know"
- Guo et al. (2017): Confidence calibration in modern neural networks
- Google TAP (Test Analytics Platform): error signature cross-branch comparison
- Launchable: diff-fault intersection A/B testing results
- Datadog CI Visibility: infrastructure correlation via APM traces
- Trunk Flaky Tests: historical flake rate modeling

**Comment by Whatsonyourmind** [comment]
The 95-100% confidence on infrastructure failures is a calibration problem. Your LLM is outputting raw logits that have never been calibrated against actual outcomes.

**The fix has two parts:**

**1. Post-hoc calibration (immediate):**
Collect a dataset of (LLM_confidence, actual_outcome) pairs from your CI history. Then fit isotonic regression:

```python
from sklearn.isotonic import IsotonicRegression
ir = IsotonicRegression(out_of_bounds='clip')
ir.fit(raw_confidences, actual_outcomes)  # actual: 1 if code caused failure, 0 if infra
calibrated = ir.predict(new_raw_confidence)
```

Isotonic regression learns a monotonic mapping from raw scores to calibrated probabilities. If your LLM says 95% but historically only 40% of 95%-confidence predictions are correct, it maps 95% -> 40%.

**2. Blast-radius prior (structural):**
Before asking the LLM anything, compute the blast radius: what fraction of tests failed simultaneously?

```
blast_radius = n_failed_tests / n_total_tests

# Infrastructure failures have high blast radius (many tests fail at once)
# Code bugs have low blast radius (specific tests fail)
P(infrastructure | blast_radius > 0.25) ≈ 0.85  // calibrate from your history

# Bayesian update:
prior_code = 1 - P(infrastructure | blast_radius)
posterior = prior_code * LLM_likelihood / evidence
```

When 60% of tests fail simultaneously, the prior that it is infrastructure is ~85% regardless of what the LLM says about the code. This structural prior prevents the LLM from confidently blaming code for infrastructure-wide failures.

**Validation:** Plot a reliability diagram (predicted confidence vs actual accuracy in decile bins). A well-calibrated system should lie on the diagonal.

**Comment by kamilpajak** [comment]
Good points — isotonic regression and Bayesian priors are where this should eventually land.

Right now we're at ~30 labeled cases, so we shipped a rule-based V1 with hard caps as an immediate fix (PR #51). We'll log `(predicted_confidence, actual_resolution)` pairs as the tool runs, and evaluate data-driven calibration once we have enough data.

Tracking that in #50 — input welcome when we get there.

**Comment by Whatsonyourmind** [comment]
The rule-based V1 with hard caps is the right call at ~30 labeled cases — isotonic regression needs at least 100-200 points to avoid overfitting the monotonic mapping. Logging the (predicted, actual) pairs now means you will have a clean calibration dataset ready when the time comes. Will keep an eye on #50.

**Comment by Whatsonyourmind** [comment]
@kamilpajak With 30 labeled cases you are right to start rule-based. The blast-radius heuristic alone should catch the majority of infrastructure false positives.

When you reach ~100 labeled cases, that is the minimum for isotonic regression to be meaningfully better than rules. At that point, the calibration pipeline is:

1. Split your 100 cases into 5 folds
2. Fit isotonic regression on 4 folds, evaluate on 1
3. Average the calibration curves across folds
4. If the average Expected Calibration Error (ECE) is < 0.10, the calibrated model is production-ready

ECE < 0.10 means the average gap between predicted confidence and actual accuracy is under 10 percentage points. Most rule-based systems have ECE around 0.25-0.35, so you will see a clear improvement when you have enough data.

The Bayesian blast-radius prior can be added on top of either approach — it is a structural correction, not a replacement for calibration.
